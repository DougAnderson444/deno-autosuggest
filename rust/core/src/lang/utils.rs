use std::collections::HashMap;


pub fn compile_utf_map(static_map: &[(&str, &str)]) -> HashMap<Vec<char>, Vec<char>> {
    let mut compiled_map = HashMap::with_capacity(static_map.len());
    for (pattern, replace) in static_map {
        compiled_map.insert(
            pattern.chars().collect(),
            replace.chars().collect()
        );
    }
    compiled_map
}
