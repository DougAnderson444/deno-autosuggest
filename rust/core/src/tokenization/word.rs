pub use crate::lang::{Lang, Chars, CharPattern, PartOfSpeech};


#[derive(PartialEq, Debug, Clone)]
pub struct Word {
    pub ix:    usize,
    pub place: (usize, usize),
    pub stem:  usize,
    pub pos:   Option<PartOfSpeech>,
    pub fin:   bool,
}


impl Word {
    pub fn new(len: usize) -> Self {
        Word {
            ix:    0,
            place: (0, len),
            stem:  len,
            pos:   None,
            fin:   true,
        }
    }

    pub fn len(&self) -> usize {
        self.place.1 - self.place.0
    }

    pub fn is_empty(&self) -> bool {
        self.place.1 == self.place.0
    }

    pub fn is_function(&self) -> bool {
        match self.pos {
            Some(PartOfSpeech::Article)     => true,
            Some(PartOfSpeech::Preposition) => true,
            Some(PartOfSpeech::Conjunction) => true,
            Some(PartOfSpeech::Particle)    => true,
            _ => false,
        }
    }

    pub fn view<'a>(&self, chars: &'a [char]) -> &'a [char] {
        &chars[self.place.0 .. self.place.1]
    }

    pub fn view_mut<'a>(&self, chars: &'a mut [char]) -> &'a mut [char] {
        &mut chars[self.place.0 .. self.place.1]
    }

    pub fn dist(&self, other: &Self) -> usize {
        let (start1, end1) = self.place;
        let (start2, end2) = other.place;
        if start1 >= end2 { return start1 - end2; }
        if start2 >= end1 { return start2 - end1; }
        panic!("Malformed words: ({}, {}), ({}, {})", start1, end1, start2, end2);
    }

    pub fn join(&self, other: &Self) -> Self {
        Word {
            ix:    self.ix,
            place: (self.place.0, other.place.1),
            stem:  other.place.0 - self.place.0 + other.stem,
            pos:   None,
            fin:   other.fin,
        }
    }
}


impl Word {
    pub fn fin(mut self, fin: bool) -> Self {
        self.fin = fin;
        self
    }

    pub fn split<'a, 'b, P: CharPattern>(&'a self, chars: &'a [char], pattern: &'b P) -> WordSplit<'a, 'b, P> {
        WordSplit { word: self, chars, pattern, ix: self.ix, offset: 0 }
    }

    pub fn strip<P: CharPattern>(&mut self, chars: &[char], pattern: &P) -> &mut Self {
        let chars = self.view(chars);
        let left  = chars.iter()
            .take_while(|&&ch| pattern.matches(ch).unwrap_or(false))
            .count();
        let right = chars.iter()
            .rev()
            .take_while(|&&ch| pattern.matches(ch).unwrap_or(false))
            .take(chars.len() - left)
            .count();
        self.place.0 += left;
        self.place.1 -= right;
        self.fin = self.fin || right != 0;
        self
    }

    pub fn stem(&mut self, chars: &[char], lang: &Lang) -> &mut Self {
        self.stem = lang.stem(self.view(chars));
        self
    }

    pub fn mark_pos(&mut self, chars: &[char], lang: &Lang) -> &mut Self {
        self.pos = lang.get_pos(self.view(chars));
        self
    }

    pub fn lower(&mut self, chars: &mut [char]) -> &mut Self {
        let chars = self.view_mut(chars);
        if chars.iter().any(|ch| ch.is_uppercase()) {
            for ch in chars {
                *ch = ch.to_lowercase().next().unwrap_or(*ch);
            }
        }
        self
    }
}


#[derive(Debug)]
pub struct WordSplit<'a, 'b, P: CharPattern> {
    word:    &'a Word,
    chars:   &'a [char],
    pattern: &'b P,
    ix:      usize,
    offset:  usize,
}


impl<'a, 'b, P: CharPattern> Iterator for WordSplit<'a, 'b, P> {
    type Item = Word;

    fn next(&mut self) -> Option<Self::Item> {
        let Self { word, ix, offset, pattern, .. } = self;
        let chars = word.view(self.chars);

        if *offset >= word.len() {
            return None;
        }

        *offset += chars[*offset ..]
            .iter()
            .take_while(|&&ch| pattern.matches(ch).unwrap_or(false))
            .count();

        let len = chars[*offset ..]
            .iter()
            .take_while(|&&ch| !pattern.matches(ch).unwrap_or(false))
            .count();

        if len == 0 {
            return None;
        }

        let splitted = Word {
            ix:     *ix,
            place:  (word.place.0 + *offset, word.place.0 + *offset + len),
            stem:   len,
            pos:    None,
            fin:    word.fin || *offset + len < word.len(),
        };

        *offset += splitted.len();
        *ix     += 1;

        Some(splitted)
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::utils::to_vec;
    use crate::lang::lang_english;
    use super::{Word, Chars, PartOfSpeech};

    use Chars::{
        Whitespaces,
        Punctuation,
    };

    #[test]
    fn word_dist_basic() {
        let mut w1 = Word::new(7);
        let mut w2 = Word::new(5);
        w1.place = (0, w1.len());
        w2.place = (w1.place.1 + 2, w1.place.1 + 2 + w2.len());
        assert_eq!(w1.dist(&w2), 2);
        assert_eq!(w2.dist(&w1), 2);
    }

    #[test]
    fn word_dist_fused() {
        let mut w1 = Word::new(7);
        let mut w2 = Word::new(5);
        w1.place = (0, w1.len());
        w2.place = (w1.place.1, w1.place.1 + w2.len());
        assert_eq!(w1.dist(&w2), 0);
        assert_eq!(w2.dist(&w1), 0);
    }

    #[test]
    #[should_panic]
    fn word_dist_malformed() {
        let mut w1 = Word::new(7);
        let mut w2 = Word::new(5);
        w1.place = (0, w1.len());
        w2.place = (w1.place.1 - 2, w1.place.1 - 2 + w2.len());
        w1.dist(&w2);
    }

    #[test]
    fn word_join_basic() {
        let mut w1 = Word::new(7);
        let mut w2 = Word::new(5);
        w1.place = (0, w1.len());
        w2.place = (w1.place.1 + 1, w1.place.1 + 1 + w2.len());
        assert_debug_snapshot!(w1.join(&w2));
    }

    #[test]
    fn word_join_offset() {
        let mut w1 = Word::new(7);
        let mut w2 = Word::new(5);
        w1.place = (3, 3 + w1.len());
        w2.place = (w1.place.1 + 1, w1.place.1 + 1 + w2.len());
        assert_debug_snapshot!(w1.join(&w2));
    }

    #[test]
    fn word_join_unfinished_first() {
        let mut w1 = Word::new(7);
        let mut w2 = Word::new(5);
        w1.place = (0, w1.len());
        w2.place = (w1.place.1 + 1, w1.place.1 + 1 + w2.len());
        w1.fin   = false;
        w2.fin   = true;
        assert_debug_snapshot!(w1.join(&w2));
    }

    #[test]
    fn word_join_unfinished_last() {
        let mut w1 = Word::new(7);
        let mut w2 = Word::new(5);
        w1.place = (0, w1.len());
        w2.place = (w1.place.1 + 1, w1.place.1 + 1 + w2.len());
        w1.fin   = true;
        w2.fin   = false;
        assert_debug_snapshot!(w1.join(&w2));
    }

    #[test]
    fn word_join_pos_first() {
        let mut w1 = Word::new(7);
        let mut w2 = Word::new(5);
        w1.place = (0, w1.len());
        w2.place = (w1.place.1 + 1, w1.place.1 + 1 + w2.len());
        w1.pos   = Some(PartOfSpeech::Article);
        w2.pos   = None;
        assert_debug_snapshot!(w1.join(&w2));
    }

    #[test]
    fn word_join_pos_last() {
        let mut w1 = Word::new(7);
        let mut w2 = Word::new(5);
        w1.place = (0, w1.len());
        w2.place = (w1.place.1 + 1, w1.place.1 + 1 + w2.len());
        w1.pos   = None;
        w2.pos   = Some(PartOfSpeech::Article);
        assert_debug_snapshot!(w1.join(&w2));
    }

    #[test]
    fn word_split() {
        let chars = to_vec(" Foo Bar, Baz; ");
        let word  = Word::new(chars.len());
        let split = word.split(&chars[..], &[Whitespaces, Punctuation]).collect::<Vec<_>>();
        assert_debug_snapshot!(split);
    }

    #[test]
    fn word_split_empty() {
        let chars = to_vec(" ,;");
        let word  = Word::new(chars.len());
        let split = word.split(&chars[..], &[Whitespaces, Punctuation]).collect::<Vec<_>>();
        assert_debug_snapshot!(split);
    }

    #[test]
    fn word_split_unfinished() {
        let chars1 = to_vec(" Foo Bar, Baz");
        let chars2 = to_vec(" Foo Bar, Baz; ");
        let word1  = Word::new(chars1.len()).fin(false);
        let word2  = Word::new(chars2.len()).fin(false);
        let split1 = word1.split(&chars1[..], &[Whitespaces, Punctuation]).collect::<Vec<_>>();
        let split2 = word2.split(&chars2[..], &[Whitespaces, Punctuation]).collect::<Vec<_>>();
        assert_eq!(split1.last().unwrap().fin, false);
        assert_eq!(split2.last().unwrap().fin, true);
    }

    #[test]
    fn word_strip() {
        let chars = to_vec(" Foo; ");
        let mut word = Word::new(chars.len());
        word.strip(&chars[..], &[Whitespaces, Punctuation]);
        assert_debug_snapshot!(&word);
    }

    #[test]
    fn word_strip_empty() {
        let chars = to_vec(" ,;");
        let mut word = Word::new(chars.len());
        word.strip(&chars[..], &[Whitespaces, Punctuation]);
        assert_debug_snapshot!(word);
    }

    #[test]
    fn word_strip_unfinished() {
        let chars1 = to_vec(" Foo Bar, Baz");
        let chars2 = to_vec(" Foo Bar, Baz; ");
        let mut word1 = Word::new(chars1.len()).fin(false);
        let mut word2 = Word::new(chars2.len()).fin(false);
        word1.strip(&chars1[..], &[Whitespaces, Punctuation]);
        word2.strip(&chars2[..], &[Whitespaces, Punctuation]);
        assert_eq!(word1.fin, false);
        assert_eq!(word2.fin, true);
    }

    #[test]
    fn word_stem() {
        let chars = to_vec("university");
        let lang  = lang_english();
        let mut word = Word::new(chars.len());
        word.stem(&chars[..], &lang);
        assert_eq!(word.stem, 7);
    }

    #[test]
    fn word_pos() {
        let lang   = lang_english();
        let chars1 = to_vec("university");
        let chars2 = to_vec("the");
        let mut word1 = Word::new(chars1.len());
        let mut word2 = Word::new(chars2.len());
        word1.mark_pos(&chars1[..], &lang);
        word2.mark_pos(&chars2[..], &lang);
        assert_eq!(word1.pos, None);
        assert_eq!(word2.pos, Some(PartOfSpeech::Article));
    }

    #[test]
    fn word_lower() {
        let mut chars = to_vec(" Foo Bar, Baz; ");
        let mut word  = Word::new(chars.len());
        word.lower(&mut chars[..]);
        assert_debug_snapshot!(word);
    }
}
