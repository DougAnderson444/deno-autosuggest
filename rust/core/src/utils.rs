#![macro_use]

use std::cell::RefCell;


macro_rules! min {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::min($x, min!($($z),*)));
}


macro_rules! max {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::max($x, max!($($z),*)));
}
