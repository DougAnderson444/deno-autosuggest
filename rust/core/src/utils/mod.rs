#![macro_use]

mod trigrams;
mod limitsort;
mod fading_windows;

pub use trigrams::{Trigrams, TrigramIter};
pub use limitsort::{LimitSort, LimitSortIter};
pub use fading_windows::FadingWindows;


macro_rules! min {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::min($x, min!($($z),*)));
}


macro_rules! max {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::max($x, max!($($z),*)));
}


pub fn to_vec<T: AsRef<str>>(chars: T) -> Vec<char> {
    chars.as_ref().chars().collect()
}


#[cfg(test)]
pub fn to_str<T: AsRef<[char]>>(chars: T) -> String {
    chars.as_ref().iter().collect()
}
