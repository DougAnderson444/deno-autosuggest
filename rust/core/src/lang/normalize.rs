use std::collections::HashMap;
use crate::utils::FadingWindows;

const NORM_MAX_PATTERN_LEN: usize = 2;


pub struct Normalize<'a> {
    windows: FadingWindows<'a, char>,
    map:     &'a HashMap<Vec<char>, Vec<char>>,
    skip:    usize,
}


impl<'a> Normalize<'a> {
    pub fn new(source: &'a [char], map: &'a HashMap<Vec<char>, Vec<char>>) -> Self {
        Self {
            windows: FadingWindows::new(source, NORM_MAX_PATTERN_LEN),
            map,
            skip: 0,
        }
    }
}


impl<'a> Iterator for Normalize<'a> {
    type Item = (&'a [char], &'a [char]);

    fn next(&mut self) -> Option<Self::Item> {
        let mut window = self.windows.next()?;

        while self.skip > 0 {
            window = self.windows.next()?;
            self.skip -= 1;
        }

        for len in (1 .. window.len() + 1).rev() {
            let pattern = &window[..len];
            if let Some(replace) = self.map.get(pattern) {
                self.skip = pattern.len() - 1;
                return Some((pattern, replace));
            }
        }

        Some((&window[..1], &window[..1]))
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        self.windows.size_hint()
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use super::super::utils::compile_utf_map;
    use super::Normalize;

    fn normalize<'a>(word: &'a [char]) -> (Vec<String>, Vec<String>) {
        let norm_map = compile_utf_map(&[
            ("ó", "o"),
            ("ó", "o"),
            ]);
        let mut source = Vec::new();
        let mut norm   = Vec::new();
        for (chunk1, chunk2) in Normalize::new(word, &norm_map) {
            source.push(chunk1.iter().collect::<String>());
            norm  .push(chunk2.iter().collect::<String>());
        }
        (source, norm)
    }

    #[test]
    fn utf_normalize_empty() {
        let input  = "".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_singleton_normal() {
        let input  = "f".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_singleton_replaced() {
        let input  = "ó".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_single_nfd_replaced() {
        let input  = "ó".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_normal() {
        let input  = "foobar".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfc_beginning() {
        let input  = "óoobar".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfd_beginning() {
        let input  = "óoobar".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfc_middle() {
        let input  = "foóbar".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfd_middle() {
        let input  = "foóbar".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfc_end() {
        let input  = "foobaó".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfd_end() {
        let input  = "foobaó".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }
}