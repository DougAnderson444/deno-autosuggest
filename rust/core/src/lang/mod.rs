mod utils;
mod normalize;
mod lang_english;
mod lang_german;
mod lang_portuguese;
mod lang_russian;
mod lang_spanish;

pub use lang_german::lang_german;
pub use lang_english::lang_english;
pub use lang_spanish::lang_spanish;
pub use lang_portuguese::lang_portuguese;
pub use lang_russian::lang_russian;

use std::cell::RefCell;
use std::collections::HashMap;
use rust_stemmers::Stemmer;
use crate::tokenization::PartOfSpeech;
use normalize::Normalize;


const BUFFER_CAPACITY: usize = 2;


pub struct Lang {
    stem_buffer:  RefCell<String>,
    norm_buffer1: RefCell<Vec<char>>,
    norm_buffer2: RefCell<Vec<char>>,
    pos_map:      HashMap<Vec<char>, PartOfSpeech>,
    compose_map:  HashMap<Vec<char>, Vec<char>>,
    reduce_map:   HashMap<Vec<char>, Vec<char>>,
    stemmer:      Stemmer,
}


impl Lang {
    pub fn new(
        pos_map:     HashMap<Vec<char>, PartOfSpeech>,
        compose_map: HashMap<Vec<char>, Vec<char>>,
        reduce_map:  HashMap<Vec<char>, Vec<char>>,
        stemmer:     Stemmer
    ) -> Self {
        let stem_buffer  = RefCell::new(String::with_capacity(BUFFER_CAPACITY));
        let norm_buffer1 = RefCell::new(Vec::with_capacity(BUFFER_CAPACITY));
        let norm_buffer2 = RefCell::new(Vec::with_capacity(BUFFER_CAPACITY));
        Self { stem_buffer, norm_buffer1, norm_buffer2, pos_map, compose_map, reduce_map, stemmer }
    }

    pub fn stem(&self, word: &[char]) -> usize {
        let buffer = &mut *self.stem_buffer.borrow_mut();
        buffer.clear();
        buffer.extend(word.iter());
        let stem = self.stemmer.stem(&buffer);
        stem.chars().count()
    }

    pub fn get_pos(&self, word: &[char]) -> Option<PartOfSpeech> {
        self.pos_map.get(word).cloned()
    }

    pub fn utf_compose(&self, word: &[char]) -> Option<Vec<char>> {
        let buffer = &mut *self.norm_buffer1.borrow_mut();
        buffer.clear();

        for (_, norm_chunk) in Normalize::new(word, &self.compose_map) {
            buffer.extend(norm_chunk);
        }

        if &buffer[..] == word {
            None
        } else {
            Some(buffer.clone())
        }
    }

    pub fn utf_reduce(&self, word: &[char]) -> Option<(Vec<char>, Vec<char>)> {
        let buffer1 = &mut *self.norm_buffer1.borrow_mut();
        let buffer2 = &mut *self.norm_buffer2.borrow_mut();
        buffer1.clear();
        buffer2.clear();

        for (word_chunk, norm_chunk) in Normalize::new(word, &self.reduce_map) {
            buffer1.extend(word_chunk);
            buffer2.extend(norm_chunk);
            for _ in 0 .. norm_chunk.len() - word_chunk.len() {
                buffer1.push('\0');
            }
        }

        if &buffer2[..] == word {
            None
        } else {
            Some((buffer1.clone(), buffer2.clone()))
        }
    }
}


#[cfg(test)]
mod tests {
    use std::collections::HashMap;
    use rust_stemmers::{Stemmer, Algorithm};
    use insta::assert_debug_snapshot;
    use super::Lang;
    use super::utils::compile_utf_map;

    fn get_lang() -> Lang {
        let stemmer  = Stemmer::create(Algorithm::English);
        let compose_map = compile_utf_map(&[("ó", "ó")]);
        let reduce_map  = compile_utf_map(&[
            ("ó", "o"),
            ("ß", "ss"),
        ]);
        let lang = Lang::new(HashMap::new(), compose_map, reduce_map, stemmer);
        lang
    }

    #[test]
    fn utf_compose_nfc() {
        let input  = "foóbar".chars().collect::<Vec<_>>();
        let output = get_lang().utf_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_compose_nfd() {
        let input  = "foóbar".chars().collect::<Vec<_>>();
        let output = get_lang().utf_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_reduce_reduced() {
        let input  = "foobar".chars().collect::<Vec<_>>();
        let output = get_lang().utf_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_reduce_nfc() {
        let input  = "foóbar".chars().collect::<Vec<_>>();
        let output = get_lang().utf_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_reduce_fill0() {
        let input  = "straße".chars().collect::<Vec<_>>();
        let output = get_lang().utf_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }
}