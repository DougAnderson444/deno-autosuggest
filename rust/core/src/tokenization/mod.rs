mod pos;
mod pattern;
mod word;
mod text;

use crate::lang::Lang;
pub use pos::PartOfSpeech;
pub use word::Word;
pub use text::Text;
pub use pattern::{Chars, CharPattern};


pub fn tokenize_query(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
    Text::from_str(source)
        .normalize(lang)
        .fin(false)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}


pub fn tokenize_record(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
     Text::from_str(source)
        .normalize(lang)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}
