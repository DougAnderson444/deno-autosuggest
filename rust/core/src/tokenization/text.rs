use std::fmt;
use crate::lang::Lang;
use super::CharPattern;
use super::Word;


#[derive(PartialEq)]
pub struct Text<T: AsRef<[char]>> {
    pub source: T,
    pub chars: T,
    pub words: Vec<Word>,
}


impl Text<Vec<char>> {
    pub fn from_vec(source: Vec<char>) -> Text<Vec<char>> {
        let len   = source.len();
        let chars = source.clone();
        Text {
            source,
            chars,
            words: vec![Word::new(len)],
        }
    }

    pub fn from_str(source: &str) -> Text<Vec<char>> {
        Text::from_vec(source.chars().collect())
    }
}


impl Text<Vec<char>> {
    pub fn to_ref<'a>(&'a self) -> Text<&'a [char]> {
        Text {
            source: &self.source,
            chars:  &self.chars,
            words:  self.words.clone(),
        }
    }
}


impl<'a> Text<&'a [char]> {
    pub fn to_own(&self) -> Text<Vec<char>> {
        Text {
            source: self.source.to_vec(),
            chars:  self.chars.to_vec(),
            words:  self.words.clone(),
        }
    }
}


impl<T: AsRef<[char]>> Text<T> {
    pub fn fin(mut self, fin: bool) -> Self {
        if let Some(word) = self.words.last_mut() {
            word.fin = fin;
        }
        self
    }

    pub fn is_empty(&self) -> bool {
        self.words.is_empty()
    }
}


impl Text<Vec<char>> {
    pub fn normalize(mut self, lang: &Option<Lang>) -> Self {
        if let Some(lang) = lang {
            if self.words.len() == 0 {
                return self;
            }
            if self.words.len() > 1 {
                panic!("Normalization should always be the first step");
            }

            if let Some(nfc) = lang.utf_compose(&self.source) {
                self.source           = nfc.clone();
                self.chars            = nfc;
                self.words[0].place.1 = self.chars.len();
            }

            if let Some((source, chars)) = lang.utf_reduce(&self.chars) {
                self.source           = source;
                self.chars            = chars;
                self.words[0].place.1 = self.chars.len();
            }
        }
        self
    }

    pub fn split<P: CharPattern>(mut self, pattern: &P) -> Self {
        let mut words = Vec::with_capacity(self.words.len());
        for word in &self.words {
            for splitted in word.split(&self.chars, pattern) {
                words.push(splitted);
            }
        }
        self.words = words;
        for (ix, word) in self.words.iter_mut().enumerate() {
            word.ix = ix;
        }
        self
    }

    pub fn strip<P: CharPattern>(mut self, pattern: &P) -> Self {
        for word in &mut self.words {
            word.strip(&self.chars, pattern);
        }
        self.words.retain(|w| w.len() > 0);
        for (ix, word) in self.words.iter_mut().enumerate() {
            word.ix = ix;
        }
        self
    }

    pub fn stem(mut self, lang: &Option<Lang>) -> Self {
        if let Some(lang) = lang {
            for word in &mut self.words {
                word.stem(&self.chars, lang);
            }
        }
        self
    }

    pub fn mark_pos(mut self, lang: &Option<Lang>) -> Self {
        if let Some(lang) = lang {
            for word in &mut self.words {
                word.mark_pos(&self.chars, lang);
            }
        }
        self
    }

    pub fn lower(mut self) -> Self {
        if self.chars.iter().any(|ch| ch.is_uppercase()) {
            for ch in &mut self.chars {
                *ch = ch.to_lowercase().next().unwrap_or(*ch);
            }
        }
        self
    }
}


impl<T: AsRef<[char]>> fmt::Debug for Text<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Text {{")?;
        for word in &self.words {
            let chars = word.view(self.chars.as_ref());
            write!(f, " \"")?;
            for (i, ch) in chars.iter().enumerate() {
                write!(f, "{}", ch)?;
                if i == word.stem - 1 && i != word.len() - 1 {
                    write!(f, "|")?;
                }
            }
            write!(f, "\"")?;
            if !word.fin { write!(f, "..")?; }
        }
        write!(f, " }}")?;
        Ok(())
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lang::{lang_english, lang_portuguese, lang_german};
    use super::{Word, Text};
    use super::super::{Chars, PartOfSpeech};

    use Chars::{
        Whitespaces,
        Punctuation,
    };

    #[test]
    fn text_normalize_nfd() {
        let text = Text::from_str("Conceição").normalize(&Some(lang_portuguese()));
        assert_debug_snapshot!((&text.source, &text.chars, &text.words[0]));
    }

    #[test]
    fn text_normalize_pad0() {
        let text = Text::from_str("straße").normalize(&Some(lang_german()));
        assert_debug_snapshot!((&text.source, &text.chars, &text.words[0]));
    }

    #[test]
    fn text_split() {
        let text = Text::from_str(" Foo Bar, Baz; ").split(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(text);
    }

    #[test]
    fn text_split_empty() {
        let text = Text::from_str(", ").split(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(text);
    }

    #[test]
    fn text_split_unfinished() {
        let text1 = Text::from_str(" Foo Bar, Baz"  ).fin(false).split(&[Whitespaces, Punctuation]);
        let text2 = Text::from_str(" Foo Bar, Baz; ").fin(false).split(&[Whitespaces, Punctuation]);
        assert_eq!(text1.words.last().unwrap().fin, false);
        assert_eq!(text2.words.last().unwrap().fin, true);
    }

    #[test]
    fn text_strip() {
        let chars = "-Foo- , Baz; ".chars().collect::<Vec<_>>();
        let text  = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words:  vec![
                    Word { ix: 0, place: (0,  5), stem: 5, pos: None, fin: true },  // "-Foo-"
                    Word { ix: 1, place: (6,  7), stem: 1, pos: None, fin: true },  // ","
                    Word { ix: 2, place: (8, 13), stem: 5, pos: None, fin: true },  // "Baz; "
                ],
            }
            .strip(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(text);
        assert_debug_snapshot!(text.words);
    }

    #[test]
    fn text_strip_unfinished() {
        let chars = "-Foo- Baz; ".chars().collect::<Vec<_>>();
        let text1 = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words:  vec![
                    Word { ix: 0, place: (0, 5), stem: 5, pos: None, fin: true },  // "-Foo-"
                    Word { ix: 1, place: (5, 8), stem: 3, pos: None, fin: true },  // "Baz"
                ],
            }
            .fin(false)
            .strip(&[Whitespaces, Punctuation]);

        let text2 = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words:  vec![
                    Word { ix: 0, place: (0,  5), stem: 5, pos: None, fin: true },  // "-Foo-"
                    Word { ix: 1, place: (5, 10), stem: 5, pos: None, fin: true },  // "Baz; "
                ],
            }
            .fin(false)
            .strip(&[Whitespaces, Punctuation]);

        assert_eq!(text1.words.last().unwrap().fin, false);
        assert_eq!(text2.words.last().unwrap().fin, true);
    }

    #[test]
    fn text_lower() {
        let chars = "Foo, Bar Baz".chars().collect::<Vec<_>>();
        let text  = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words:  vec![
                    Word { ix: 0, place: (0,  4), stem: 4, pos: None, fin: true }, // "Foo,"
                    Word { ix: 1, place: (5,  8), stem: 3, pos: None, fin: true }, // "Bar"
                    Word { ix: 2, place: (9, 12), stem: 3, pos: None, fin: true }, // "Baz"
                ],
            }
            .lower();
        assert_debug_snapshot!(text);
    }

    #[test]
    fn text_stem() {
        let chars = "hello universe".chars().collect::<Vec<_>>();
        let lang  = lang_english();
        let text  = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words: vec![
                    Word { ix: 0, place: (0,  5), stem: 5, pos: None, fin: true }, // "hello"
                    Word { ix: 1, place: (6, 14), stem: 8, pos: None, fin: true }, // "universe"
                ],
            }.stem(&Some(lang));
        assert_eq!(text.words[0].stem, 5);
        assert_eq!(text.words[1].stem, 7);
    }

    #[test]
    fn text_pos() {
        let chars = "the universe".chars().collect::<Vec<_>>();
        let lang  = lang_english();
        let text  = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words: vec![
                    Word { ix: 0, place: (0,  3), stem: 3, pos: None, fin: true }, // "hello"
                    Word { ix: 1, place: (4, 12), stem: 8, pos: None, fin: true }, // "universe"
                ],
            }.mark_pos(&Some(lang));
        assert_eq!(text.words[0].pos, Some(PartOfSpeech::Article));
        assert_eq!(text.words[1].pos, None);
    }
}
