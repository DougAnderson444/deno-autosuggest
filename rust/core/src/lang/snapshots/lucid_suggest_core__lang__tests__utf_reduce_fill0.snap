---
source: src/lang/mod.rs
expression: output
---
Some(
    (
        [
            's',
            't',
            'r',
            'a',
            'ÃŸ',
            '\u{0}',
            'e',
        ],
        [
            's',
            't',
            'r',
            'a',
            's',
            's',
            'e',
        ],
    ),
)
